Caltech-IPAC/firefly
# FIREFLY-71: MultiProductViewer component

This component allows us to take in a table that represents any type of data product and show the product. A sample implementation uses obs core but it is made for any type of data source.

  - refactor to ImageMetaDataWatcher.
  - Created new DataProductsFactory
  - improved some image return feedback and messaging
  - Now supports menu of data products
  - support multiple product types
  - support both tables, png, images, and download

Much of this work is really positioning firefly for products. There is still some issues that are not yet done and need future tickets:

- chart support, this is easy but I have not chart test data.
- the table handling the 'more...' dropdown. 
- the obscore converter could be worked on significantly more to get better at recognizing data but I did enough to demonstrate the potential.

Next steps:

- implement a data products converter for Sophia and 
- implement a data products converter Spitizer (if we do the convertions)
- Improve and generalize the data product converter for Datalink.

How it works in a nutshell:
`DataProductsFactory` recognizes a table type and produces a converter for that table. The converter will produce an object that knows how to load a data type using an activate function that it creates. It can also produce a menu of other data that the row represents. The `MultiProductViewer` will change to the data type of call an `activate` function that will do the data load. It will also show the menu of other options.

Key files to look at.

- `MultiProductsViewer.jsx`
- `ImageMetaDataWatcher.js`
- `DataProductsFactory.js`

_To test:_

https://irsawebdev9.ipac.caltech.edu/firefly-71-data-products/firefly/firefly-dev.html

Do an obscore search with the left hand side should show various data products.  Also with CADC you will see a `More...` menu that will allow you to switch your data products.
